Fix redirect vulnerability

oliverguenther opened the pull request:
`redirect_back_or_default` was vulnerable to some of the URLs found to
be vulnerable in redmine, such as `@test.foo`.

This extends the method with a path check to avoid these cases.
Thanks to @marutosi for pointing this out.

http://www.redmine.org/projects/redmine/repository/revisions/14560

oliverguenther committed:
Fix redirect vulnerability

`redirect_back_or_default` was vulnerable to some of the URLs found to
be vulnerable in redmine, such as `@test.foo`.

This commit extracts the whole functionality into a policy and alters
the constraints with a path check to avoid these cases.
Thanks to @marutosi for pointing this out.

http://www.redmine.org/projects/redmine/repository/revisions/14560

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-redirect-uri.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-redirect-uri.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented on app/controllers/application_controller.rb:
```diff
@@ -450,6 +450,12 @@ def redirect_back_or_default(default, escape = true, use_escaped = true)
         uri = URI.parse(escaped_back_url)
 
+        # Remove basic auth credentials
+        uri.userinfo = nil
```
This is not covered by a spec @oliverguenther. BTW, why do you think it necessary to remove it?

oliverguenther replied:
The idea stems from the original fix, but one possible attack vector in our case would be guessing API authentication (global user credentials). 

Do we have a legitimate use case for basic auth on the application itself? It may be possible that someone uses, e.g., basic auth around an OP page, but then again the basic auth is cached by the browser itself, isn't it?

ulferts replied:
I didn't think about installations not having SSL enabled. I always assume that SSL is active but that of course is wishful thinking. So I am OK with removing the basic auth credentials (even though the attacker could have gotten the credentials on the request easily)

Plus I didn't know that the browser caches basic auth (OMG). I learned that from you, thanks.

Spec is still necessary, then.

oliverguenther replied:
It's a bit of a vague vector, as an attacker would still need to have some process to gather the returned status code in order to succesfully test credentials. 

I still think my reasoning is correct in that: We remove it and IF there is a use-case that we do not foresee, then we can still whitelist it again. I'm not clear why anyone would redirect to a basic auth, though.

Caching btw. differs from browser to browser, but I think all major vendors do that by default.

ulferts commented:
We should probably take the chance to refactor the method before merging release/4.2 into dev. It doesn't have to be within this PR, though.

oliverguenther commented:
> We should probably take the chance to refactor the method before merging release/4.2 into dev. It doesn't have to be within this PR, though.

I'll see what I can do, it is unlikely that we'll touch it again in a new PR, thus let's rather do it here.

oliverguenther commented:
@ulferts I've just pushed the refactored version, which extracts the check into a separate policy.

basiszwo commented on app/policies/redirect_policy.rb:
```diff
@@ -0,0 +1,151 @@
+  # - Tries to parse it
+  # - Escapes the redirect URL when requested so.
+  def preprocess(requested, escape)
+    url = if escape
```
write as one-liner? `url = escape ? URI.escape(CGI.unescape(requested.to_s)) : requested`

basiszwo commented on app/controllers/application_controller.rb:
```diff
@@ -439,57 +439,14 @@ def back_url
-      # TODO explain reasoning for this
-      account/register
-      )}x # ignore whitespace
+    policy = RedirectPolicy.new(
```
shrink to one line?

oliverguenther commented:
@ulferts , @basiszwo I've addressed the broken omniauth spec in 4.2, this should be good to go.

NobodysNightmare commented on app/policies/redirect_policy.rb:
```diff
@@ -0,0 +1,148 @@
+  ##
+  # Avoid paths with references to parent paths
+  def no_upper_levels
+    !@requested_url.path.include? '..'
```
1. I am not sure if there is an attack vector from using `..` back URLs
   - I could specify any path using an absolute path `/work_packages`
   - I can't traverse above the root level of my "directory" structure, because our paths do not map into the fileystem: `community.openproject.org/../../../etc/passwd`
     - even if I could, this attack should (also) be caught at another level
2. this check is too rigid, `..foo` and `foo..bar`` are valid file or directory names that would not pass
3. there are no specs for this validation

oliverguenther replied:
If you use a relative root directory, you could bypass the relative check and redirect to other subdirectories by using `/mysubdir/../evildir/foobar`.

NobodysNightmare replied:
aha, that makes sense (and again I read your comment too late ^^)

This would only leave my points two and three (this was was edited in) valid...

oliverguenther replied:
Note that this exact check is in place in the code before: `!escaped_back_url.match(/\.\./)`

The comment regarding specs is valid however. We could probably expand the URL to absolute instead of checking for two dots. I'll look into it.

NobodysNightmare replied:
I actually had a brief look at the old code to check whether this was introduced by you or present before.
So good job on making the validation more obvious!

Making the URl absolute sounds like a good solution...

NobodysNightmare commented on app/policies/redirect_policy.rb:
```diff
@@ -0,0 +1,148 @@
+      # normally cause you to be redirected to this page. As it is the logout page, however,
+      # this would log you right out again after a successful login.
+      logout |
+      # TODO explain reasoning for this
```
TODO ^^

edit: apparently an older todo... -.-

oliverguenther replied:
I could give no better explanation than 'do not redirect to register', thus I left that remark as-is.

oliverguenther committed:
Always preprocess URLs with CGI.unescape

This forces redirect URLs to always pass through the escaping process,
in order to process escaped paths (`/mysubdir/%2E%2E/secret/`).

oliverguenther commented:
@NobodysNightmare I've addressed your remarks, note that I check for `../` for sake of simplicity, and it's a small step forward from the existing check in 4.2.

redirect URLs are now always escaped and unescaped for URIs, thus turning `2%E2%E/` correctly into `../`.

NobodysNightmare commented:
One step after another they say ^^

LGTM, waiting for travis to agree :horse:
